polars/polars-io/src/cloud/adaptors.rs

```diff
@@ -129,3 +132,166 @@ impl AsyncSeek for CloudReader {
         std::task::Poll::Ready(Ok(self.pos))
     }
 }
+
+/// Adaptor which wraps the asynchronous interface of [ObjectStore::put_multipart](https://docs.rs/object_store/latest/object_store/trait.ObjectStore.html#tymethod.put_multipart)
+/// exposing a synchronous interface which implements `std::io::Write`.
+///
+/// This allows it to be used in sync code which would otherwise write to a simple File or byte stream,
+/// such as with `polars::prelude::CsvWriter`.
+pub struct CloudWriter {
+    // Hold a reference to the store
+    object_store: Arc<dyn ObjectStore>,
+    // The path in the object_store which we want to write to
+    path: Path,
+    // ID of a partially-done upload, used to abort the upload on error
+    multipart_id: MultipartId,
+    // The Tokio runtime which the writer uses internally.
+    runtime: tokio::runtime::Runtime,
+    // Internal writer, constructed at creation
+    writer: std::sync::Mutex<Box<dyn AsyncWrite + Send + Unpin>>,
```

Comment:
Why does this need to be behind a mutex? Don't we have `&mut self` when we write?

Reply:
This is a good question, and one I wanted to ask myself.

The conondrum is: Do we want `CloudWriter` as exposed to public users to be `Sync`?
If it is not, any writer wrapping `CloudWriter` is also not `Sync`.

The trade-off is that requiring users to deal with this is making their life more complicated, while on the other hand a Mutex lock is not a free abstraction that they'd pay for even if they don't use multiple threads for writing.

Simple usage of `CloudWriter` (like `CsvWriter<CloudWriter>`) does not need to be  `Sync` since it will be used from a single thread. But code using e.g. `ParquetWriter` that wants to write multiple morsels, like own usage inside the `ParquetCloudSink`, require a `Sync` bound.


What do you prefer? Moving the mutex to the implementation of `ParquetCloudSink`, or leaving it here?

Reply:
> If it is not, any writer wrapping CloudWriter is also not Sync.

A user can always make it sync by putting `CloudWriter` in a `Mutex`/ `Arc<Mutex<>>`. We don't need the mutex, so let's drop it.